1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+
+`ICustomAstVisitor2` and `AstVisitor2` would be extended to deal with this AST,
+and .NET Core 3's new default interface implementation feature would be
+leveraged to ensure this does not break things
+as previous syntactic introductions have been forced to.
+
+#### Statements may not be chained
+
+PowerShell has a notable divergence from POSIX shells:
+
+- In POSIX shells, a pipeline is composed of statements
+- In PowerShell, a pipeline is a kind of statement
+
+This means that bash allows the following:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF'
+```
+
+In fact, this can be evaluated in the background:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF' &
+```
+
+However, in PowerShell, pipelines are subordinate to statements.
+The proposed implementation of chain operators precludes use of
+`&&` and `||` between statements like `if` or `while`,
+or constructions like `$x = cmd1 && $y = cmd2 && $x + $y`.
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+## Related Material
+
+- Original issue: [PowerShell/PowerShell #3241](https://github.com/PowerShell/PowerShell/issues/3241).
+
+- Work-in-progress implementation: [PowerShell/PowerShell #9849](https://github.com/PowerShell/PowerShell/pull/9849).
+
+- [Current handling of `&&` and `||` in the parser](https://github.com/PowerShell/PowerShell/blob/af1de9e88d28014438ff3414e82298e5b14f6e81/src/System.Management.Automation/engine/parser/Parser.cs#L5846-L5860).
+
+## Alternate Proposals and Considerations
+
+### `&&` and `||` as a statement separator
+
+To be more bash-like, an alternative implementation might treat
+`&&` and `||` with the same precedence as `;`, as a way to separate statements.
+
+This would allow assignment within chains, like:
+
+```powershell
+$x = cmd1 && $y = cmd2 && $x + $y
+```
+
+Other possibilities would include:
+
+```powershell
+if ($condition) { cmd1 } && while ($anotherCondition) { cmd2 }
+```
+
+```powershell
+foreach ($v in 1..100)
+{
+    cmd1 $v && break
+}
+```
+
+#### Background operator changes
+
+In bash, background operators are lower precedence than chain separators
+(`&` has the same precedence as `;`).
+In PowerShell, background operators are higher precedence.
+
+Making chain operators apply to any statement would lead to the question
+of the precedence of the background operator.
+
+The example:
+
+```powershell
+cmd1 && cmd2 &
+```
+
+Could either be `[[cmd1 && cmd2] &]` or `[cmd1 && [cmd2 &]]`.
+
+Having chain operators apply to statements would mean the second case
+is the correct one, unless changes are made to the background operator precedence (which would also allow `if ($condition) { cmd1 } &)` for example).
+
+The use of this is possibly less than the use of being able to background
+an entire pipeline chain; `cmd1 & && cmd2` would be equivalent to
+`cmd1 &; cmd2` since a backgrounded pipeline
+will never fail in the invoking context.

```

Comment:
Right, that's why my default proposal is to background the entire chain. Evaluating the success of `cmd1 &` is not helpful and possibly confusing.

Being able to do `cmd1 && cmd2 &` seems much more useful to me.

Reply:
Bear with me while I'm think something out loud here. I've had asynchronous execution of PowerShell on my mind lately, and I'm exploring those thoughts a bit.

What if you changed the way `cmd1 &` is evaluated for success? Sure, a job gets created, which is always successful unless there's some odd bug in `Start-Job`, so that's pretty useless. How would you really want to measure the success of a job? Wouldn't it be better to measure whether that job completes without error or not?

PowerShell could benefit from a lot better asynchronous support, so I find myself wondering if this syntax:

```powershell
cmd1 & && cmd2 & && cmd3 & || cmd4
```

Could be interpreted as a way to launch `cmd1` in the background, and if that job succeeds, launch `cmd2` in the background, and if that job succeeds launch `cmd3` in the background, and if any of them fail invoke cmd4 locally, with the results of this command being a task (i.e. what if when the `&` background operator is used in a pipeline chain, the result is automatically a task, because it only makes sense in that scenario for the entire chain to be run asynchronously). Or if we don't like tasks for this, I suppose I could use this syntax to get a job instead:

```powershell
. {cmd1 & && cmd2 & && cmd3 & || cmd4} &
```

Like I said, I'm just thinking out loud, but if the `&` background operator gets `$!` support for the current job that is running (see some discussion in [PowerShell Issue 9873](https://github.com/PowerShell/PowerShell/issues/9873#issuecomment-501289658), then `cmd4` could use that to identify which job failed and perform error handling for that job.

The flipside of this is that if `&` has lower precedence than `&&` and `||` as is proposed here, to accomplish the same you would have to write something like this:

```powershell
# But how would PowerShell recognize that this should return a task, since the background operators are embedded in script blocks? That doesn't seem to work.
.{cmd1 &} && .{cmd2 &} && .{cmd3 &} || cmd4
```

or

```powershell
# This is explicit, but much more cumbersome syntactically
.{.{cmd1 &} && .{cmd2 &} && .{cmd3 &} || cmd4} &
```

So it may still be doable, but with a bunch of extra enclosures to change the precedence.

I do think it's a scenario worth thinking about though, and I really like the notion that a pipeline chain with a background operator at the end of one of the pipelines would result in the entire chain being run in the background automatically, with the measurement of success for the segment that is specifically backgrounded being whether or not the job itself had errors. I think that gives you what you are looking for (easy backgrounding of a chain) while supporting more powerful scenarios where you want to daisy-chain multiple background jobs while handling failures in a DRY way.

Reply:
The more I think about my previous comment, the more I really like it. Other thoughts/comments would be appreciated.

Reply:
Could we move this interesting proposal in another RFC to keep current RFC more simple?

Reply:
Some of this belongs in another RFC. That's already done. See here: #193.

The rest cannot be moved to another RFC. If I did that, it would be a competing RFC because otherwise it would incur breaking changes if one was implemented, and then the other.

We need to make the right decision on precedence now, not in a separate RFC.

We should also agree that with chaining operators, measuring job success requires more than just checking if the job was launched. It makes much more sense and aligns with the bash precedence, which I believe that community will expect.

Reply:
> with the entire chain running as a job

In the case why do we use "||" in `cmd1 & && cmd2 & && cmd3 & || cmd4`?
I'd expect that '|' is a job operator too:
```powershell
cmd1 & && cmd2 & && cmd3 | cmd4
```

Reply:
@iSazonov Note that I went back into my comment above (the thinking out loud comment) and cleaned it up a bit so that it's less confusing and more clearly demonstrates what I hope to see come out of this.

Reply:
> In the case why do we use "||" in `cmd1 & && cmd2 & && cmd3 & || cmd4`?

@iSazonov `||` tells PowerShell to run the following segment in the chain only if the last run segment failed (which may be `cmd1 &` or `cmd2 &` or `cmd3 &`). That's how chaining works. I'm not sure I understand the expectation that `|` is a job operator. Also note that as I mentioned, I cleaned up my thoughts, so the specific quote you referenced isn't there anymore and I think it makes much more sense now (please re-read).

Reply:
I see you point. My thoughts is that we could simplify syntax and introduce two job operators '&' and '|' symmetrically to expression operators '&&' and '||' 

```powershell
cmd1 & cmd2 & cmd3 | cmd4
```
where cmd1, cmd2 and cmd3 run in parallel and cmd4 runs if any of previous failed.

Reply:
Except that `|` is a pipeline separator, so that couldn't possibly work.

Reply:
This does not exclude the possibility of interpreting it differently in this context. If of course it is possible to identify this context.

Reply:
I still think an additional context for a single `|` couldn't possibly work. In `cmd1 & cmd2 & cmd3 | cmd4`, I read that as "run `cmd1` and `cmd2` in the background and `cmd3 | cmd4` in the current runspace.

Reply:
I think you're off in the weeds:

What is the value of `cmd1 & && cmd2` vs `cmd1 && cmd2`?
Nothing.

**UNLIKE** other programming languages, you get no benefit from `await` because your thread doesn't have anything better to do than wait around for `cmd1` to finish.

Reply:
@Jaykul I'm working with jobs, and looking at this as a way to wait for them and then proceed based on the results. That's what I meant by "like having `await`".

What is the value of `cmd1 & cmd2 & cmd3 & || cmd4`? Launch `cmd1`, `cmd2`, and `cmd3` asynchronously in the background, and if any of them fails, launch `cmd4`. You get benefit from waiting for those jobs to complete, and only launching `cmd4` if one or more of the jobs comes back with an error.

Also, the argument I'm making is that I disagree with the proposed operator precedence, which suggests that `&` should have a higher precedence than `&&` and `||`. This is the opposite of the precedence in bash, where we're borrowing these operators from (that alone raises a red flag). By having `&&` and `||` have a higher precedence than `&`, and properly checking to see if a job succeeded or failed, which means waiting on the job(s) (where I was finding similarity with `await`) to check the results instead of just checking to see if the job launched successfully or not (which is useless), we get some additional fan-out support in PowerShell.

Reply:
@Jaykul Also you're thinking code-only, not ad-hoc/shell. The value of `cmd1 & && cmd2 &` in a shell, in particular if some of RFC #193 gets accepted/implemented such that jobs can be launched and still show visible progress without blocking the current shell, is very high when compared to `cmd1 && cmd2`.